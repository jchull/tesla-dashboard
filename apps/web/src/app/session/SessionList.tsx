import React, { useLayoutEffect, useRef, useState } from 'react'
import './session.style.scss'

import { useDispatch, useSelector } from 'react-redux'

import { SessionListItem } from './SessionListItem'
import { fetchSessionListAction, selectSession, SessionState } from './actions'
import { AppState } from '../store'

export const SessionList: React.FC<SessionState> = (props) => {
  const dispatch = useDispatch()
  const scroll = useRef<HTMLDivElement>(null)
  const loading = useSelector((store: AppState) => store.session.loading)
  const loadedCount = useSelector(
    (store: AppState) => store.session.loadedCount
  )
  const totalCount = useSelector((store: AppState) => store.session.totalCount)
  const selectedProductId = useSelector(
    (store: AppState) => store.product.selectedProductId
  )

  const [lastRequestTime, setLastRequestTime] = useState(Date.now()
                                                             .valueOf())

  useLayoutEffect(() => {
    scroll.current.addEventListener('scroll', onScroll)
    return () => scroll.current.removeEventListener('scroll', onScroll)
  })

  const onScroll = (event: Event) => {
    if (
      !loading &&
      loadedCount < totalCount &&
      Date.now()
          .valueOf() - lastRequestTime > 1000 &&
      // @ts-ignore
      event.target.scrollHeight -
      // @ts-ignore
      event.target.clientHeight -
      // @ts-ignore
      event.target.scrollTop -
      100 <
      0
    ) {
      setLastRequestTime(Date.now()
                             .valueOf())
      dispatch(
        fetchSessionListAction(selectedProductId, {
          start: loadedCount,
          size: 100
        })
      )
    }
  }

  return (
    <div className='session-list'>
      <div className='sessions'
           ref={scroll}>
        {props.sessions.length
          ? props.sessions.map((session) => (
            <SessionListItem
              session={session}
              selected={session._id === props.selectedSessionId}
              selectionHandler={() => dispatch(selectSession(session._id))}
              key={session._id}
            />
          ))
          : null}
      </div>
    </div>
  )
}
