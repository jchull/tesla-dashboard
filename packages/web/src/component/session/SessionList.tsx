import React from 'react';
import './session.style.scss';

import { useDispatch } from 'react-redux';

import { SessionListItem } from './SessionListItem';
import { selectSession, SessionState } from './actions';
import { FilterToolbar } from '@teslapp/web/src/component/toolbar/FilterToolbar';

export const SessionList: React.FC<SessionState> = (props: SessionState) => {
  const dispatch = useDispatch();

  return (
    <div className="session-list">
      <div className="sessions">
        {// @ts-ignore
        props.sessions &&
          props.sessions.map((session) => (
            <SessionListItem
              session={session}
              selected={session._id === props.selectedSessionId}
              selectionHandler={() => dispatch(selectSession(session._id))}
              key={session._id}
            />
          ))}
      </div>
      <FilterToolbar />
    </div>
  );
};
