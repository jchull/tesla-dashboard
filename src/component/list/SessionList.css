.session-list {
    width: 320px;
    display: inline-block;
    vertical-align: top;
    margin: 10px 0 10px 20px;
    height: 100%;
}

.selected-view {
    display: inline-block;
    min-height: 100%;
    overflow-y: auto;
}


h4 {
    margin: 12px;
    padding: 10px 10px 5px 10px;
    text-transform: uppercase;
    background: linear-gradient(20deg, rgba(51, 51, 51, 0.0), rgba(0,0,20, 0.2));
    color: #c00;
    box-shadow: 1px 2px 2px #222222;
    text-shadow: 1px 2px 2px rgba(0,0,0,0.2);
    letter-spacing: 1px;
}
