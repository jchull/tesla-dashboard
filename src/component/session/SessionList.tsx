import React from 'react';
import {useDispatch} from 'react-redux';

import {SessionListItem} from './SessionListItem';
import {selectSession, SessionState} from './actions';
import './style.scss';


export const SessionList: React.FC<SessionState> = (props) => {

  const dispatch = useDispatch();

  return (
      <div className="session-list">
        {
          props.sessions.map(session =>
                                 <SessionListItem session={session}
                                                  selected={session._id === props.selectedSessionId}
                                                  selectionHandler={() => dispatch(selectSession(session._id))}
                                                  key={session._id}/>)
        }
      </div>
  );
};

