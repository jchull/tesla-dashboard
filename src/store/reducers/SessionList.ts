import {ACTION_TYPES, SessionListAction} from '../actions';
import {IVehicleSession} from 'tesla-dashboard-api';

const initialState = {
  sessionList: [] as IVehicleSession[]
};

// TODO: SessionAction should be part of a union type
export function sessionListReducer(state = initialState, action: SessionListAction) {
  switch (action.type) {
    case ACTION_TYPES.DELETE_SESSION:
      return {
        sessionList: state.sessionList.filter((session: IVehicleSession) => action.sessionId !== session._id)
      };
      break;
    case ACTION_TYPES.ARCHIVE_SESSION:
      return {
        sessionList: state.sessionList.filter((session: IVehicleSession) => action.sessionId !== session._id)
      };
      break;
      // case ACTION_TYPES.UPDATE_SESSION_LIST:
      //   break;
      // case ACTION_TYPES.FILTER_SESSION_LIST:
      //   break;
    default:
      return state;
  }
}