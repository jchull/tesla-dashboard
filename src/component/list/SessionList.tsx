import React from 'react';
import './SessionList.scss';
import {SessionListItem} from './SessionListItem';
import {queryService} from '@service/Services';
import {isDriveSession} from '../../type/util';
import {LineChart} from '@component/chart/LineChart';
import {SessionTagList} from '@component/vehicle/SessionTagList';
import {IVehicleSession, IVehicleState} from 'tesla-dashboard-api';
import {SessionListState} from '@store/types/state';
import {ChartToolbar} from '@component/chart/ChartToolbar';


export const SessionList: React.FC<SessionListState> = props => {
  const [selectedSession, setSelectedSession] = React.useState({} as IVehicleSession);
  const [selectedDatum, setSelectedDatum] = React.useState([] as IVehicleState[]);

  const sessionSelectionHandler = (session: IVehicleSession) => setSelectedSession(Object.assign({}, session));

  React.useEffect(() => {
    if (selectedSession._id && props.vehicle) {
      if (isDriveSession(selectedSession)) {
        queryService.getDrivingStates(props.vehicle.vin, selectedSession._id)
                    .then(setSelectedDatum);
      } else {
        queryService.getChargingStates(props.vehicle.vin, selectedSession._id)
                    .then(setSelectedDatum);
      }
    }

  }, [selectedSession]);

  return (
      <div className="session-list-container">
        <div className="session-list">
          {
            props.sessionList.map(session =>
                                      <SessionListItem session={session}
                                                       selected={session._id === selectedSession._id}
                                                       selectionHandler={() => sessionSelectionHandler(session)}
                                                       key={session._id}/>)
          }
        </div>
        {selectedSession && selectedDatum && selectedDatum.length > 0 && props.vehicle ?
         <div className="selected-view">
           <LineChart vehicle={props.vehicle}
                      session={selectedSession}
                      states={selectedDatum}/>
           <SessionTagList vehicleId={props.vehicle.vin}
                           sessionId={selectedSession._id}
                           tags={selectedSession.tags}/>
           <ChartToolbar product={props.vehicle} sessionId={selectedSession._id}/>

         </div>
                                                                                       :
         <div className="selected-view message">
           Nothing selected
         </div>
        }

      </div>
  );
};

