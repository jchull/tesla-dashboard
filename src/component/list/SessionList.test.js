import React                              from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act }                            from 'react-dom/test-utils';
import { SessionList }                    from './SessionList';

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


it('renders empty session list', () => {
  act(() => {
    render(<SessionList sessions={[]}/>, container);
  });
  // expect(container.innerHTML)
  //   .toMatchInlineSnapshot();
});

