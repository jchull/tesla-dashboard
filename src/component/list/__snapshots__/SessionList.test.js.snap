// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty session list 1`] = `
"<div class=\\"session-list-container\\">
  <div class=\\"session-list\\"></div>
  <div class=\\"selected-view message\\">No product selected</div>
</div>"
`;
